### ISSUE TITLE
FaradayMiddleware::FollowRedirects loops forever on Location: redirect with a url-encoded #url-fragment

### ISSUE THREAD

**Issue by jm3** [lostisland/faraday_middleware#149]
_NOTE: I originally reported this to Typhoeus, but was @iMacTia [thinks this is likely an issue](https://github.com/lostisland/faraday/issues/566#issuecomment-269656257) a **faraday_middleware** issue_

This URL redirects forever back-and-forth to the server until the `FaradayMiddleware::FollowRedirects` limit is hit: **http://whathappensinadops.tumblr.com/post/48847516848/when-theres-something-on-the-io-that-we-dont-even**
### log:

```
: get http://whathappensinadops.tumblr.com/post/48847516848/when-theres-something-on-the-io-that-we-dont-even
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont-even%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
: get http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont%23_=_
Status: 301
ruby-2.2.3/gems/faraday_middleware-0.10.0/lib/faraday_middleware/response/follow_redirects.rb:80:
in `block in perform_with_redirection': 
too many redirects; last one to: 
http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont#_=_ 
(FaradayMiddleware::RedirectLimitReached)

 ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/response.rb:65:in `call'
 ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/response.rb:65:in `block in finish'
 ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/response.rb:65:in `each'
 ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/response.rb:65:in `finish'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/adapters/faraday.rb:100:in `block in request'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/request/callbacks.rb:128:in `call'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/request/callbacks.rb:128:in `block in execute_callbacks'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/request/callbacks.rb:127:in `each'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/request/callbacks.rb:127:in `execute_callbacks'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/request/operations.rb:35:in `finish'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/easy_factory.rb:129:in `block in set_callback'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/easy/response_callbacks.rb:65:in `call'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/easy/response_callbacks.rb:65:in `block in complete'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/easy/response_callbacks.rb:65:in `each'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/easy/response_callbacks.rb:65:in `complete'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/multi/operations.rb:151:in `check'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/multi/operations.rb:164:in `run'
 ruby-2.2.3/gems/ethon-0.8.0/lib/ethon/multi/operations.rb:43:in `perform'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/hydra/runnable.rb:15:in `run'
 ruby-2.2.3/gems/typhoeus-0.8.0/lib/typhoeus/hydra/memoizable.rb:51:in `run'
 ruby-2.2.3/gems/faraday-0.9.2/lib/faraday/connection.rb:278:in `in_parallel'
 ./fragment-redirect-overflow.rb:15:in `<main>'
```

**Note** that the trailing `#url-fragment` is added by the server in the [first] Location redirect (as a url-encoded %23), it's not in the original url, so this is at least partly a Tumblr bug. That said, it works fine in both `curl` and Chrome with after following a single Location: redirect, not 50, so I would sort of expect Faraday to behave a little more like that.

Should Faraday be url-decoding the %23 and then discarding the #fragment?

This is also reproducible with this Tumblr URL: http://mtvinsights.tumblr.com/post/32884529556/50-shades-of-nerd; seems to be what happens when a tumblr user renames their blog — Tumblr serves a 301 redirect response with a trailing #_=_ that sends Typhoeus / Faraday into an infinite loop.

## Typhoeus repro:
see [this gist](https://gist.github.com/jm3/14469eb935dc8974f531)

_@iMacTia reports that the same happens if you try to do it with the :net_http adapter:_
## net_http repro:

    conn = Faraday.new('http://whathappensinadops.tumblr.com') do |f|
      f.use FaradayMiddleware::FollowRedirects
      f.adapter :net_http
    end
    
    conn.get('/post/48847516848/when-theres-something-on-the-io-that-we-dont-even')
    # FaradayMiddleware::RedirectLimitReached: too many redirects; last one to: 
    # http://happensinadops.com/post/48847516848/when-theres-something-on-the-io-that-we-dont#_=_

**Comment by iMacTia** [comment]
Thanks @jm3 for reporting this, marked as bug.
Will try to fix and release with 0.10.2 after having a careful read at the [RFC Standard](https://tools.ietf.org/html/rfc7231#section-6.4)

**Comment by iMacTia** [comment]
Ok, I think the [RFC Section 7.1.2](https://tools.ietf.org/html/rfc7231#section-7.1.2) is very clear on this.
Location headers may contain a fragment component and it needs to be managed accordingly.

**Comment by iMacTia** [comment]
And I just found other issues related to the same thing (without the infinite loop issue though): #138, #117

**Comment by iMacTia** [comment]
@jm3 I've opened #150 
I would like to know if you can give it a quick shot and let me know if it solves your issue.
Decision was to ignore the fragment component completely as suggested on https://github.com/lostisland/faraday_middleware/issues/117#issuecomment-122384169

**Comment by jm3** [comment]
Cool, I replied to #150 with my results — looks good!

I have a gem-wizardry that arose when testing this: is it because of the branch name that I can't point directly to the PR branch from my Gemfile? I tried `…, :branch => ‘#149/redirect-url-fragment’`, but it errors with **fatal: Needed a single revision**

    Revision #149/redirect-url-fragment does not exist in the repository github.com/lostisland/far….
    Maybe you misspelled it?

But if I point directly to 6e04ea6, the SHA of the only commit on that branch, it works fine. Is that because the branch name contains `#`  and `/` chars?
